### PR TITLE
Added GitHub token functionality

### PR DESCRIPTION
**Pull request by VictorLeach96**
This PR adds GitHub token support through the GitHub API.
This allows for prebuild-install to use private repos.

Tests need to be updated still

### CODE REVIEW

**Comment by vweevers** [comment]
At first glance, this makes all downloads (private or public) use the GitHub API, correct? Which means they are subject to rate limits.

**Comment by VictorLeach96** [comment]
@vweevers yep I've tweaked it now so it will only use the API when a token is available

File: asset.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+var noop = Object.assign({
+  http: function () {},
+  silly: function () {}
+}, require('noop-logger'))
+var util = require('./util')
+var url = require('url')
+var tunnel = require('tunnel-agent')
+
+function findAssetId (opts, cb) {
+  if (!opts.token) return cb()
```
Comment (ralphtheninja) [inline asset.js]: I'd prefer if we could move this outside the call, it's basically a "hidden" if-statement that also is calling back syncronously. E.g. first check if `opts.token` is set and then call `findAssetid(opts, cb)`.

File: asset.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    log.http(res.statusCode, apiUrl)
+    if (res.statusCode !== 200) return cb(err)
+
+    // Find asset id in release
+    for (var release of data) {
+      if (release.tag_name === 'v' + opts.pkg.version) {
+        for (var asset of release.assets) {
+          if (asset.browser_download_url === downloadUrl) {
+            opts.assetId = asset.id
+            return cb()
```
Comment (ralphtheninja) [inline asset.js]: `findAssetId()` should probably callback with `(null, assetId)` instead of modifying `opts` inside the function.

File: bin.js [hunk 3]
```diff
@@ -51,16 +51,27 @@ if (!isNpm && /node_modules/.test(process.cwd())) {
...
       return process.exit(1)
     }
     log.info('install', 'Successfully installed prebuilt binary!')
   })
-})
+}
+
+if (opts.token){
+  var downloadUrl = util.getAssetUrl(opts)
+  asset(opts, function (err) {
```
Comment (ralphtheninja) [inline bin.js]: `(err, assetId)`

File: asset.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  var log = opts.log || noop
+
+  log.http('request', 'GET ' + apiUrl)
+  var reqOpts = { url: apiUrl, json: true }
+  var proxy = opts['https-proxy'] || opts.proxy
+
+  reqOpts.headers = {
+    'User-Agent': 'simple-get'
+  }
+  if (opts.token) {
```
Comment (ralphtheninja) [inline asset.js]: This if statement can be removed since we know `opts.token` is set.

**Comment by ralphtheninja** [comment]
Unit tests is always possible.

**Comment by VictorLeach96** [comment]
Could use a http mock to give dummy responses?

**Comment by ralphtheninja** [comment]
> Could use a http mock to give dummy responses?

Sure. We do that in `download-test.js` https://github.com/prebuild/prebuild-install/blob/master/test/download-test.js#L131-L135

**Comment by mathiask88** [comment]
@VictorLeach96 @ralphtheninja Maybe you can outsource the proxy stuff into a function? It is nearly two times the same code?!

**Comment by ralphtheninja** [comment]
> @VictorLeach96 @ralphtheninja Maybe you can outsource the proxy stuff into a function? It is nearly two times the same code?!

Good catch. And a small test for that function!

**Comment by ralphtheninja** [comment]
So at the moment I feel there are two things missing:

- [ ] An example in `README` explaining in more detail what `--token` does and how it's used
- [ ] Refactor the proxy code

Then we can squash this PR and release it.

**Comment by VictorLeach96** [comment]
Thanks @mathiask88 @ralphtheninja should be all done now. Let me know if you happy with everything.

File: test/proxy-test.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  var request = proxy(reqOpts, opts)
+
+  t.equal(request.url, reqOpts.url, 'Request url remains the same')
+  t.equal(request.json, reqOpts.json, 'Request json remains the same')
+  t.equal(request.headers['User-Agent'], reqOpts.headers['User-Agent'], 'Request user agent remains the same')
+  t.equal(request.Authorization, reqOpts.Authorization, 'Request auth remains the same')
+
+  t.equal(request.agent.proxyOptions.host, 'hostname.com', 'Proxy hostname is set')
+  t.equal(request.agent.proxyOptions.port, 8080, 'Proxy hostname is set')
+  t.equal(request.agent.proxyOptions.proxyAuth, 'user:pass', 'Proxy hostname is set')
```
Comment (mathiask88) [inline test/proxy-test.js]: Descriptions

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "prebuild-install",
-  "version": "2.5.3",
+  "name": "prebuild-install-token",
```
Comment (mathiask88) [inline package.json]: :)

Comment (VictorLeach96) [reply package.json]: Whoops 😂

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
...
+  "version": "2.5.4",
   "description": "A command line tool for easily install prebuilds for multiple version of node/iojs on a specific platform",
```
Comment (ralphtheninja) [inline package.json]: Also this ^

**Comment by VictorLeach96** [comment]
@ralphtheninja that was already fixed. :)

**Comment by VictorLeach96** [comment]
Amazing :) thanks for all your feedback @ralphtheninja @mathiask88 !

**Comment by ralphtheninja** [comment]
@VictorLeach96 Great work! Released in 3.0.0

**Comment by mathiask88** [comment]
Cool, thanks for your effort! 🎉

**Pull request merged**